Update requestActiveNowPlayingSessionInfo to return more info

mahruskazi opened the pull request:
#### 455a582a079c81614455deebf125487e4e3f182b
<pre>
Update requestActiveNowPlayingSessionInfo to return more info
<a href="https://bugs.webkit.org/show_bug.cgi?id=269022">https://bugs.webkit.org/show_bug.cgi?id=269022</a>
<a href="https://rdar.apple.com/122427266">rdar://122427266</a>

Reviewed by Jean-Yves Avenard.

Currently requestActiveNowPlayingSessionInfo returns hasActiveSession, title, duration, elapsedTime, uniqueIdentifier and registeredAsNowPlaying.
NowPlayingInfo has other attributes that can be used, changing this function to return the struct instead.

* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::getNowPlayingInfo const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.h:
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getNowPlayingMediaTitleAndArtist:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _requestActiveNowPlayingSessionInfo:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::requestActiveNowPlayingSessionInfo):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getNowPlayingMediaTitleAndArtist):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::requestActiveNowPlayingSessionInfo):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/274432@main">https://commits.webkit.org/274432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/10fbc258e90189099fddddd20239582f32e71ae9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17694 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41108 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41294 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34416 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41070 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20510 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15043 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32519 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14898 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33655 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12936 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12909 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42570 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35196 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34916 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38741 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13573 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36951 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15185 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8744 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/14863 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -69,6 +69,7 @@
 #include <WebCore/LayerHostingContextIdentifier.h>
 #include <WebCore/MediaControlsContextMenuItem.h>
 #include <WebCore/MediaKeySystemRequest.h>
+#include <WebCore/NowPlayingInfo.h>
```
use forward declaration in the header when you can.
Only include the header in code, speeds up compilation time

Especially as you're also `#import` the header already

jyavenard commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -149,24 +150,18 @@
-        elapsedTime = sharedManager->lastUpdatedNowPlayingElapsedTime();
-        uniqueIdentifier = sharedManager->lastUpdatedNowPlayingInfoUniqueIdentifier().toUInt64();
-        registeredAsNowPlayingApplication = sharedManager->registeredAsNowPlayingApplication();
+        std::optional<NowPlayingInfo> nowPlayingInfo = sharedManager->getNowPlayingInfo();
```
`auto`

jyavenard commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2556,6 +2556,12 @@ - (void)_togglePictureInPicture
 #endif
 }
 
+- (void)_getNowPlayingMediaTitleAndArtist:(void (^)(NSString *, NSString *))completionHandler
```
`_nowPlayingMediaTitleAndArtist`

we don't use `get`

jyavenard commented on Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
```diff
@@ -89,6 +90,7 @@ class PlatformMediaSessionManager
     bool activeAudioSessionRequired() const;
     bool canProduceAudio() const;
 
+    virtual std::optional<NowPlayingInfo> getNowPlayingInfo() const { return { }; }
```
`nowPlayingInfo()`

jyavenard commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2556,6 +2556,12 @@ - (void)_togglePictureInPicture
+- (void)_getNowPlayingMediaTitleAndArtist:(void (^)(NSString *, NSString *))completionHandler
+{
+    THROW_IF_SUSPENDED;
+    _impl->getNowPlayingMediaTitleAndArtist(completionHandler);
```
ditto

jyavenard commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -375,6 +375,8 @@ for this property.
 - (void)_resumeAllMediaPlayback;
 - (void)_closeAllMediaPresentations;
 
+- (void)_getNowPlayingMediaTitleAndArtist:(void (^)(NSString *, NSString *))completionHandler;
```
ditto

jyavenard commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -221,8 +222,13 @@ - (void)_requestActiveNowPlayingSessionInfo:(void(^)(BOOL, BOOL, NSString*, doub
-        handler(active, registeredAsNowPlayingApplication, title, duration, elapsedTime, uniqueIdentifier);
+    _page->requestActiveNowPlayingSessionInfo([handler = makeBlockPtr(callback)] (bool registeredAsNowPlayingApplication, WebCore::NowPlayingInfo&& nowPlayingInfo) {
+        BOOL isActive = nowPlayingInfo.allowsNowPlayingControlsVisibility;
+        NSString *title = nowPlayingInfo.title;
```
nowPlayingInfo.title is a `String`, not a NSString

why create temporary variable, just pass the values from the structure directly to the completion handler.

jyavenard commented on Source/WebKit/UIProcess/mac/WebViewImpl.h:
```diff
@@ -649,6 +649,8 @@ ALLOW_DEPRECATED_DECLARATIONS_END
 
     AVTouchBarScrubber *mediaPlaybackControlsView() const;
 #endif
+    void getNowPlayingMediaTitleAndArtist(void(^completionHandler)(NSString *, NSString *));
```
```suggestion
    void nowPlayingMediaTitleAndArtist(void(^completionHandler)(NSString *, NSString *));
```

jyavenard commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -149,24 +150,18 @@
-        uniqueIdentifier = sharedManager->lastUpdatedNowPlayingInfoUniqueIdentifier().toUInt64();
-        registeredAsNowPlayingApplication = sharedManager->registeredAsNowPlayingApplication();
+        std::optional<NowPlayingInfo> nowPlayingInfo = sharedManager->getNowPlayingInfo();
+        if (nowPlayingInfo.has_value()) {
```
```suggestion
        if (nowPlayingInfo) {
```

jyavenard commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -149,24 +150,18 @@
+        std::optional<NowPlayingInfo> nowPlayingInfo = sharedManager->getNowPlayingInfo();
+        if (nowPlayingInfo.has_value()) {
+            bool registeredAsNowPlayingApplication = sharedManager->registeredAsNowPlayingApplication();
+            completionHandler(registeredAsNowPlayingApplication, WTFMove(nowPlayingInfo.value()));
```
```suggestion
            completionHandler(registeredAsNowPlayingApplication, WTFMove(*nowPlayingInfo));
```

mahruskazi committed:
Update requestActiveNowPlayingSessionInfo to return more info
https://bugs.webkit.org/show_bug.cgi?id=269022
rdar://122427266

Reviewed by Jean-Yves Avenard.

Currently requestActiveNowPlayingSessionInfo returns hasActiveSession, title, duration, elapsedTime, uniqueIdentifier and registeredAsNowPlaying.
NowPlayingInfo has other attributes that can be used, changing this function to return the struct instead.

* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
(WebCore::PlatformMediaSessionManager::getNowPlayingInfo const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.h:
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getNowPlayingMediaTitleAndArtist:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _requestActiveNowPlayingSessionInfo:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::requestActiveNowPlayingSessionInfo):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::getNowPlayingMediaTitleAndArtist):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::requestActiveNowPlayingSessionInfo):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/274432@main
